File: go/deliberation/pkg/deliberation/specialists/empiricalprior/empiricalprior.go
// Package empiricalprior — Go port of @accretive-substrate/deliberation/specialists/empirical-prior.
//
// Mirrors coach.mjs::applyEmpiricalPriors: reads 24h forward-return
// distribution for the signal pattern; n>=5 floor; WR<35%+median<-0.5%
// → sell, WR>65%+median>0.5% → buy, else hold.
package empiricalprior

import (
	"fmt"

	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation"
)

const (
	Name           = "empirical-prior"
	MinN           = 5
	BearWRMax      = 35.0
	BearMedianMax  = -0.5
	BullWRMin      = 65.0
	BullMedianMin  = 0.5
)

func Argue(pkg deliberation.EvidencePackage) deliberation.Voice {
	priors := pkg.EmpiricalPriors
	if priors == nil || priors.H24 == nil {
		return deliberation.AbstainBecause(Name, "no 24h prior in evidence")
	}
	h24 := priors.H24

	citations := []deliberation.Citation{{
		Type:    "empirical_prior",
		Horizon: "24h",
		N:       h24.N,
		WR:      h24.WinRatePct,
		Median:  h24.MedianPct,
		Pattern: priors.Pattern,
	}}

	if h24.N < MinN {
		return deliberation.AbstainBecause(Name,
			fmt.Sprintf("n=%d below %d sample floor", h24.N, MinN))
	}

	if h24.WinRatePct < BearWRMax && h24.MedianPct < BearMedianMax {
		conf := 0.65 + (BearWRMax-h24.WinRatePct)/100 + abs(h24.MedianPct)/5
		if conf > 0.92 {
			conf = 0.92
		}
		return deliberation.VoteFor(Name, deliberation.VerdictSell, conf,
			fmt.Sprintf("empirical bear: n=%d, 24h WR=%.0f%%, median=%.2f%%",
				h24.N, h24.WinRatePct, h24.MedianPct),
			citations)
	}

	if h24.WinRatePct > BullWRMin && h24.MedianPct > BullMedianMin {
		conf := 0.65 + (h24.WinRatePct-BullWRMin)/100 + h24.MedianPct/5
		if conf > 0.92 {
			conf = 0.92
		}
		return deliberation.VoteFor(Name, deliberation.VerdictBuy, conf,
			fmt.Sprintf("empirical bull: n=%d, 24h WR=%.0f%%, median=%.2f%%",
				h24.N, h24.WinRatePct, h24.MedianPct),
			citations)
	}

	return deliberation.VoteFor(Name, deliberation.VerdictHold, 0.5,
		fmt.Sprintf("pattern n=%d, 24h WR=%.0f%%, median=%.2f%% — no decisive edge",
			h24.N, h24.WinRatePct, h24.MedianPct),
		citations)
}

func abs(x float64) float64 {
	if x < 0 {
		return -x
	}
	return x
}

// Specialist — exported safe-wrapped argue function.
var Specialist = deliberation.MakeSafe(Name, Argue)

File: node/deliberation/src/specialists/empirical-prior.mjs
/**
 * @accretive-substrate/deliberation/specialists/empirical-prior
 *
 * Empirical-prior specialist. Reads forward-return distributions for the
 * current signal-source pattern from pkg.empirical_priors and emits a
 * verdict based on historical win-rate + median return at the 24h horizon.
 *
 * Data source: scripts/source-conflict-empirical.py output, the same
 * priors file coach.mjs::applyEmpiricalPriors() already consumes. The
 * orchestrator's assemble step (B2-future) will pre-compute the matching
 * pattern signature.
 *
 * Verdict rules (deterministic, mirrors coach's empirical-prior logic):
 *   pattern n >= 5, 24h WR < 35%, median < -0.5%   → sell (strong bear)
 *   pattern n >= 5, 24h WR > 65%, median >  0.5%   → buy  (strong bull)
 *   pattern n < 5                                  → abstain (insufficient)
 *   else                                           → hold  (no edge)
 *
 * Domain: 'all' — empirical priors apply to every venue.
 */

import { makeSpecialist, voteFor, abstainBecause } from './base.mjs';

export const EMPIRICAL_PRIOR_SPECIALIST_NAME = 'empirical-prior';
const MIN_N = 5;
const BEAR_WR_MAX = 35;
const BEAR_MEDIAN_MAX = -0.5;
const BULL_WR_MIN = 65;
const BULL_MEDIAN_MIN = 0.5;

async function argue(pkg) {
  const priors = pkg.empirical_priors;
  if (!priors || !priors.h24) {
    return abstainBecause(EMPIRICAL_PRIOR_SPECIALIST_NAME, 'no 24h prior in evidence');
  }
  const h24 = priors.h24;
  const n = Number(h24.n) || 0;
  const wr = Number(h24.win_rate_pct);
  const median = Number(h24.median_pct);

  const citations = [
    { type: 'empirical_prior', horizon: '24h', n, wr, median, pattern: priors.pattern || null },
  ];

  if (n < MIN_N) {
    return abstainBecause(
      EMPIRICAL_PRIOR_SPECIALIST_NAME,
      `n=${n} below ${MIN_N} sample floor`,
    );
  }

  if (wr < BEAR_WR_MAX && median < BEAR_MEDIAN_MAX) {
    const confidence = Math.min(0.92, 0.65 + (BEAR_WR_MAX - wr) / 100 + Math.abs(median) / 5);
    return voteFor(
      EMPIRICAL_PRIOR_SPECIALIST_NAME, 'sell', confidence,
      `empirical bear: n=${n}, 24h WR=${wr.toFixed(0)}%, median=${median.toFixed(2)}%`,
      citations,
    );
  }

  if (wr > BULL_WR_MIN && median > BULL_MEDIAN_MIN) {
    const confidence = Math.min(0.92, 0.65 + (wr - BULL_WR_MIN) / 100 + median / 5);
    return voteFor(
      EMPIRICAL_PRIOR_SPECIALIST_NAME, 'buy', confidence,
      `empirical bull: n=${n}, 24h WR=${wr.toFixed(0)}%, median=${median.toFixed(2)}%`,
      citations,
    );
  }

  return voteFor(
    EMPIRICAL_PRIOR_SPECIALIST_NAME, 'hold', 0.5,
    `pattern n=${n}, 24h WR=${wr.toFixed(0)}%, median=${median.toFixed(2)}% — no decisive edge`,
    citations,
  );
}

export const empiricalPriorSpecialist = makeSpecialist(
  EMPIRICAL_PRIOR_SPECIALIST_NAME,
  ['all'],
  argue,
);
